sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -3038,6 +3025,107 @@ class Dataset[T] private[sql](
     dropDuplicates(colNames)
   }
 
+  /**
+   * Returns a new Dataset with duplicates rows removed, within watermark.
+   *
+   * For a static batch [[Dataset]], it just drops duplicate rows. For a streaming [[Dataset]],
+   * this will keep all data across triggers as intermediate state to drop duplicated rows. The
+   * state will be kept to guarantee the semantic, "Events are deduplicated as long as the time
+   * distance of earliest and latest events are smaller than the delay threshold of watermark."
+   * The watermark for the input [[Dataset]] must be set via [[withWatermark]]. Users are
+   * encouraged to set the delay threshold of watermark longer than max timestamp differences
+   * among duplicated events. In addition, too late data older than watermark will be dropped.
+   *
+   * @group typedrel
+   * @since 3.5.0
+   */
+  def dropDuplicatesWithinWatermark(): Dataset[T] = {
+    dropDuplicatesWithinWatermark(this.columns)
+  }
+
+  /**
+   * Returns a new Dataset with duplicates rows removed, considering only the subset of columns,
+   * within watermark.
+   *
+   * For a static batch [[Dataset]], it just drops duplicate rows. For a streaming [[Dataset]],
+   * this will keep all data across triggers as intermediate state to drop duplicated rows. The
+   * state will be kept to guarantee the semantic, "Events are deduplicated as long as the time
+   * distance of earliest and latest events are smaller than the delay threshold of watermark."
+   * The watermark for the input [[Dataset]] must be set via [[withWatermark]]. Users are
+   * encouraged to set the delay threshold of watermark longer than max timestamp differences
+   * among duplicated events. In addition, too late data older than watermark will be dropped.
+   *
+   * @group typedrel
+   * @since 3.5.0
+   */
+  def dropDuplicatesWithinWatermark(colNames: Seq[String]): Dataset[T] = {
```

Comment:
Should we maybe have `dropDuplicatesWithinWatermark(cols: String*)` one alone? Uses can invoke `dropDuplicatesWithinWatermark(Seq(...): _*)`. That will automatically create `dropDuplicatesWithinWatermark(colNames: Array[String])` signature too for Java IIRC

Reply:
I'm not sure Scala users would feel convenient to use API which require `: _*` to be set by themselves.

I wouldn't like users to have different UX than existing dropDuplicates as they are sibling APIs. I just tried to add some cases for Scala and Java, and unfortunately removing any of methods don't go well.

Reply:
E.g. Your suggestion passes the new test addition in sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java but had to change (`: _*` as you mentioned) Scala test suite to pass. Having the same set of overloaded methods makes all tests pass. (Here I meant able to compile as pass.)

Reply:
Just to clarify, having one method would allow the cases below:

```python
def dropDuplicatesWithinWatermark(cols: String*)
```

In Scala:

```scala
df.dropDuplicatesWithinWatermark(Seq(col1, col2): _*)
df.dropDuplicatesWithinWatermark(Array(col1, col2): _*)
df.dropDuplicatesWithinWatermark(col1, col2, ...)
```

In Java

```java
String[] arr = {co1, col2}
df.dropDuplicatesWithinWatermark(arr)
df.dropDuplicatesWithinWatermark(col1, col2, ...)
```

The current signatures in the PR:

```scala
def dropDuplicatesWithinWatermark(colNames: Seq[String])
def dropDuplicatesWithinWatermark(colNames: Array[String])
```

adds two more Scala use cases:

```scala
df.dropDuplicatesWithinWatermark(Seq(col1, col2))
df.dropDuplicatesWithinWatermark(Array(col1, col2))
```

I guess using `:_*` is pretty common given that `Dataset.select` or `Dataset.selectExpr` do not have such variants too.